src/Toml/Parser.hs

```diff
@@ -79,12 +80,66 @@ basicStringP :: Parser Text
 basicStringP = lexeme $ Text.pack <$> (char '"' *> anyChar `manyTill` char '"')
 
 textP :: Parser Text
-textP = literalStringP <|> basicStringP
+textP = multilineStringP <|> multilineLiteralStringP <|> literalStringP <|> basicStringP
 
 -- adds " or ' to both sides
 quote :: Text -> Text -> Text
 quote q t = q <> t <> q
 
+nonControlChar :: Parser Text
+nonControlChar = Text.singleton <$> satisfy (not . isControl)
+
+multilineStringP :: Parser Text
+multilineStringP = lexeme $ fmap mconcat $ quotes3
+                                        >> optional eol
+                                        >> allowedChar `manyTill` quotes3
+  where
+    quotes3 = string "\"\"\""
+
+    allowedChar :: Parser Text
+    allowedChar = lineEndingBackslash <|> escapeSequence <|> nonControlChar <|> eol
+
+    lineEndingBackslash :: Parser Text
+    lineEndingBackslash = Text.empty <$ try (char '\\' >> eol >> space)
+
+    escapeSequence :: Parser Text
+    escapeSequence = char '\\' >> anyChar
+                 >>= \case
+                      'b'  -> return "\b"
+                      't'  -> return "\t"
+                      'n'  -> return "\n"
+                      'f'  -> return "\f"
+                      'r'  -> return "\r"
+                      '"'  -> return "\""
+                      '\\' -> return "\\"
+                      'u'  -> parseHexUnicode 4
+                      'U'  -> parseHexUnicode 8
+                      _    -> fail ""
+      where
+        parseHexUnicode :: Int -> Parser Text
+        parseHexUnicode n = count n hexDigitChar
+                        >>= \x -> case (toUnicode . toInt) x of
+                              Just c  -> return (Text.singleton c)
+                              Nothing -> fail ""
+          where
+            toInt xs = (read :: String -> Int) ("0x" ++ xs)
```

Comment:
We already have parser for hexadecimal number. I think we can reuse it here. If it's too complicated, then at least it's better to enable paranoia mode and use `readMaybe` function.

Reply:
Also, it's better to write top-level type singatures for functions inside `where` instead of specifying types internally. So this is better:

```haskell
toInt :: String -> Int
toInt xs = ...
```

Reply:
The parser for hexadecimal numbers that we have is not able to parse a specific number of digits (in this case we need to parser either 4 or 8 digits), so I implemented this `toInt` function.

The `read` or `readMaybe` functions will read an `Int` in decimal representation if it's not prefixed with `0x` or `0X`.

What I need here is a function that works like `read`, but it should treat the string as a hexadecimal `Int`, even if it's not prefixed with `0x`, that's why I implemented `toInt`.

Can you suggest something else that will do the job?

Reply:
@ghallak Yeah, I see that we can't reuse `hexP` parser here. It's completely okay.

Okay, usage of `read` should be completely safe here. Let's not enable paranoia mode then. So you can write this function like this:

```haskell
hexToInt :: String -> Int
hexToInt xs = read $ "0x" ++ xs
```